### PR TITLE
mojave install r openmp fix

### PR DESCRIPTION
**Pull request by orenov**
This PR is related to https://github.com/microsoft/LightGBM/issues/2026

### CODE REVIEW

**Comment by Laurae2** [comment]
ping @jameslamb

**Comment by StrikerRUS** [comment]
Maybe also support MacPorts for the consistency with Python installation script?
https://github.com/microsoft/LightGBM/blob/abbbbd7545e6a0549c17a182b7a59d7468cfb58e/python-package/setup.py#L192-L195

Won't it break the installation with gcc?
For Python we perform the following check:
https://github.com/microsoft/LightGBM/blob/abbbbd7545e6a0549c17a182b7a59d7468cfb58e/python-package/setup.py#L168-L170

**Comment by orenov** [comment]
> Maybe also support MacPorts for the consistency with Python installation script?
> 
> https://github.com/microsoft/LightGBM/blob/abbbbd7545e6a0549c17a182b7a59d7468cfb58e/python-package/setup.py#L192-L195

I see what you mean. There are some restructuring of install Rscript required, so I propose to do it as a separate pr.
> 
> Won't it break the installation with gcc?
> For Python we perform the following check:
> 
> https://github.com/microsoft/LightGBM/blob/abbbbd7545e6a0549c17a182b7a59d7468cfb58e/python-package/setup.py#L168-L170

This one I've changed. Can't make it exactly as python `if` statement, because it seems like I cannot build LightGBM R-package without OpenMP, right?

**Comment by StrikerRUS** [comment]
> I see what you mean. There are some restructuring of install Rscript required, so I propose to do it as a separate pr. (if it doesn't make any problems ofc)

Agree!
> This one I've changed. Can't make it exactly as python if statement, because it seems like I cannot build LightGBM R-package without OpenMP, right?

Thanks!

@Laurae2 @jameslamb Is `USE_OPENMP` option exposed to R-package?
https://github.com/microsoft/LightGBM/blob/abbbbd7545e6a0549c17a182b7a59d7468cfb58e/CMakeLists.txt#L10

File: R-package/src/install.libs.R [hunk 1]
```diff
@@ -44,6 +44,14 @@ if (!use_precompile) {
     cmake_cmd <- paste0(cmake_cmd, " -DUSE_R35=ON ")
   }
 
+  if (Sys.info()['sysname'] == 'Darwin' & !(grepl('^gcc', Sys.getenv('CC', '')) & grepl('^g++'), Sys.getenv('CXX', ''))) {
```
Comment (jameslamb) [inline R-package/src/install.libs.R]: could you please add a comment above this block explaining its intent?

Also, could you please use `&&` so we get short-circuiting? If you've never seen that before in R, try this:

```
> FALSE & stop("error")
Error: error

> FALSE && stop("error")
[1] FALSE
```

**Comment by jameslamb** [comment]
@StrikerRUS `USE_OPENMP` is not currently exposed in the R package but I don't see any reason it couldn't be (similar to [how we expose `USE_R35` flag](https://github.com/microsoft/LightGBM/blob/master/R-package/src/install.libs.R#L44)).

@orenov is exposing this flag to `install.libs.R` the "restructuring of install Rscript" you are referring to?

Parts of `lightgbm` specific to the R API do use OpenMP directives (e.g. [this line](https://github.com/microsoft/LightGBM/blob/master/src/lightgbm_R.cpp#L120)), but we could still expose the flag and those directives would just be ignored at compile time, right? Sorry, I am not very familiar with OpenMP.

**Comment by StrikerRUS** [comment]
@jameslamb 
>Parts of lightgbm specific to the R API do use OpenMP directives ...

I think there should be no difference compared with other cpp code...
https://github.com/microsoft/LightGBM/blob/abbbbd7545e6a0549c17a182b7a59d7468cfb58e/include/LightGBM/utils/openmp_wrapper.h#L61-L67

File: R-package/src/install.libs.R (hunk 1 above)
Comment (orenov) [reply R-package/src/install.libs.R]: Thanks for a good example.

**Comment by orenov** [comment]
> @StrikerRUS `USE_OPENMP` is not currently exposed in the R package but I don't see any reason it couldn't be (similar to [how we expose `USE_R35` flag](https://github.com/microsoft/LightGBM/blob/master/R-package/src/install.libs.R#L44)).
> 
I can't figure out how a user can pass `USE_OPENMP=OFF` to R install script. If you can give me a hint about how it's done, I can easily add it. The only thing I see is just to add next lines of code which are similar to `use_gpu`.

```
use_openmp <- TRUE
...
if (!use_openmp) {
    cmake_cmd <- paste0(cmake_cmd, " -DUSE_OPENMP=OFF ")
}
....
if (Sys.info()['sysname'] == 'Darwin' && use_openmp ... )
```
The problem for me, again,  I see this confusing as I'm not aware how user during installation can pass to install scipt `use_openmp=FALSE`. Any ideas?

> @orenov is exposing this flag to `install.libs.R` the "restructuring of install Rscript" you are referring to?
> 
Not actually. So the change which was proposed by @StrikerRUS required running cmake command a few times in the worst case. Different cases are guesses of usual OpenMP location on Mac OS. It needs to keep track of cmake status command and if it's not successful rerun cmake with different parameters.

**Comment by Laurae2** [comment]
@orenov You can pass and get those arguments using `commandArgs(trailingOnly=TRUE)` in R.

In the long term we might want to change to this instead of using hardcoded stuff in the installation script.

@StrikerRUS not yet. The user can modify itself the installation script when required to disable it. I think this is what we do currently for GPU.

**Comment by StrikerRUS** [comment]
I think that then MacPorts support and switching to `commandArgs` are content for another PR(s).

**Comment by jameslamb** [comment]
Agree with the comments from @Laurae2 and @StrikerRUS . We right now do not have the ability to pass through installation flags using `R CMD INSTALL`, and instead user has to manually change `install.libs.R`.

We'd be grateful if you open an issue to track the (separate) work of changing the install script.

**Comment by StrikerRUS** [comment]
@Laurae2 @jameslamb 
Can this PR be merged?

**Comment by jameslamb** [comment]
I am comfortable with, but don't want to merge without @Laurae2 's approval (since I'm not as familiar with OpenMP)

**Comment by attamica** [comment]
This is a life saver. I have tried everything but they all failed until I came across this change this PR suggests...Finally got the package installed successfully.

**Comment by StrikerRUS** [comment]
ping @Laurae2

**Pull request merged**